### PR TITLE
CAFF 53 - Account for loss of visual odometry features in ekf localization

### PR DESCRIPTION
**Pull request by 1999michael**
Changes:

- Extend map to not lose vis odom features when facing outside map
- Reset vis odom to last recorded value when lost (with extended map, it shouldn't be lost for any long period of time)

### CODE REVIEW

File: robot/description/launch/zed_emulation.launch [hunk 1]
```diff
@@ -5,35 +5,36 @@
...
 
-            <remap from="depth"             to="depth/depth_registered"/>
+            <remap from="depth"                 to="depth/depth_registered"/>
 
-            <param name="frame_id"          type="string" value="caffeine/base_link"/>
-            <param name="odom_frame_id"     value="caffeine/odom"/>
-            <param name="approx_sync"       value="false"/>
-            <param name="queue_size"        value="5"/>
-            <param name="publish_tf"        value="false"/>
+            <param name="Odom/ResetCountdown"   value="1"/>
```
Comment (erickmu1) [inline robot/description/launch/zed_emulation.launch:32]: A comment explaining (pointing to documentation) on its important would be good, bc it helped you resolve the visual odom getting lost issue

File: worlds/gazebo_worlds/igvc_ramp.world [hunk 2]
```diff
@@ -113,6 +113,11 @@
               <size>4.5 2 0.1</size>
             </box>
           </geometry>
+          <material>
+            <ambient>0.8 1 0.8 1</ambient>
+            <specular>0.2 0.2 0.2 64</specular>
+            <diffuse>0.4 0.4 0.4 1</diffuse>
+          </material>
```
Comment (erickmu1) [inline worlds/gazebo_worlds/igvc_ramp.world:120]: why is the only world you add a material to?

**Review by erickmu1 — CHANGES_REQUESTED** [review]
So I realized that you just expanded the plane size as your fix. Unfortunately this approach is not sufficient for making the IGVC map "larger". In particular, if you look in Gazebo, the lanes and the lane markings do not line up (the map is now stretched). **Note:** textures will always stretch to the aspect ratio of the mesh, e.g. applying a square texture on a rectangle will stretch the texture so it fully fits the rectangle (hence warping the texture). The proper way to expand the map would be to either:
1) Add rectangles (ground planes) surrounding the IGVC ground plane and then there u can just apply the grass textures
2) Add a larger plane beneath (just slightly under) the IGVC ground plane and apply a grass texture to it

That is, you need to always make the IGVC ground plane `50 x 50` and add more planes if you want to "make it bigger"

File: worlds/gazebo_worlds/igvc_ramp.world (hunk 2 above)
Comment (1999michael) [reply worlds/gazebo_worlds/igvc_ramp.world:120]: It was white before for some reason...

File: robot/description/launch/zed_emulation.launch (hunk 1 above)
Comment (1999michael) [reply robot/description/launch/zed_emulation.launch]: done

Comment (1999michael) [reply robot/description/launch/zed_emulation.launch]: done

File: worlds/gazebo_worlds/igvc_ramp.world (hunk 2 above)
Comment (erickmu1) [reply worlds/gazebo_worlds/igvc_ramp.world:120]: so this material is for colouring the ramp correct?

Comment (1999michael) [reply worlds/gazebo_worlds/igvc_ramp.world:120]: yes (it matches the other numbers)

**Review by erickmu1 — APPROVED** [review]

**Pull request merged**